cpp/src/arrow/util/decimal.cc

```diff
@@ -195,42 +191,89 @@ double Decimal128::ToDouble(int32_t scale) const {
   return DecimalDoubleConversion::ToReal(*this, scale);
 }
 
-std::string Decimal128::ToIntegerString() const {
-  Decimal128 remainder;
-  std::stringstream buf;
-  bool need_fill = false;
-
-  // get anything above 10 ** 36 and print it
-  Decimal128 top;
-  std::tie(top, remainder) = Divide(kTenTo36).ValueOrDie();
-
-  if (top != 0) {
-    buf << static_cast<int64_t>(top);
-    remainder.Abs();
-    need_fill = true;
+template <size_t n>
+static void AppendLittleEndianArrayToString(const std::array<uint64_t, n>& array,
+                                            std::string* result) {
+  static_assert(n > 0, "Array size must be positive");
+  size_t most_significant_elem_idx = n - 1;
+  while (array[most_significant_elem_idx] == 0) {
+    if (most_significant_elem_idx == 0) {
+      result->push_back('0');
+      return;
+    }
+    --most_significant_elem_idx;
+  }
+
+  std::array<uint64_t, n> copy = array;
+  constexpr uint32_t k1e9 = 1000000000U;
+  constexpr size_t num_bits = n * 64;
+  // Each segment holds at most 9 decimal digits.
+  // The number of segments needed = ceil(num_bits * log(2) / log(1e9))
+  // = ceil(num_bits / 29.897352854) <= ceil(num_bits / 29).
+  std::array<uint32_t, (num_bits + 28) / 29> segments;
+  size_t num_segments = 0;
+  uint64_t* most_significant_elem = &copy[most_significant_elem_idx];
+  do {
+    // Compute remainder = copy % 1e9 and copy = copy / 1e9.
+    uint32_t remainder = 0;
+    uint64_t* elem = most_significant_elem;
+    do {
+      uint32_t hi = static_cast<uint32_t>(*elem >> 32);
```

Comment:
Please add a comment (or make the loop its own method with a descriptive name) indicating what this is doing in little bit more detail, it took me a while to grok it.

Reply:
Done.